pulp_container/app/registry.py

```diff
@@ -81,30 +82,48 @@ def _base_paths(path):
         return [path]
 
     @staticmethod
-    async def _dispatch(path, headers):
+    async def _dispatch(file, headers):
         """
         Stream a file back to the client.
 
         Stream the bits.
 
         Args:
-            path (str): The fully qualified path to the file to be served.
-            headers (dict):
+            file (:class:`django.db.models.fields.files.FieldFile`): File to respond with
+            headers (dict): A dictionary of response headers.
+
+        Raises:
+            :class:`aiohttp.web_exceptions.HTTPFound`: When we need to redirect to the file
+            NotImplementedError: If file is stored in a file storage we can't handle
 
         Returns:
-            StreamingHttpResponse: Stream the requested content.
+            The :class:`aiohttp.web.FileResponse` for the file.
 
         """
         full_headers = MultiDict()
 
         full_headers['Content-Type'] = headers['Content-Type']
         full_headers['Docker-Content-Digest'] = headers['Docker-Content-Digest']
         full_headers['Docker-Distribution-API-Version'] = 'registry/2.0'
-        full_headers['Content-Length'] = os.path.getsize(path)
+        full_headers['Content-Length'] = str(file.size)
         full_headers['Content-Disposition'] = 'attachment; filename={n}'.format(
-            n=os.path.basename(path))
-        file_response = web.FileResponse(path, headers=full_headers)
-        return file_response
+            n=os.path.basename(file.name))
+
+        if settings.DEFAULT_FILE_STORAGE == 'pulpcore.app.models.storage.FileSystem':
+            path = os.path.join(settings.MEDIA_ROOT, file.name)
+            file_response = web.FileResponse(path, headers=full_headers)
+            return file_response
+        elif settings.DEFAULT_FILE_STORAGE == 'storages.backends.s3boto3.S3Boto3Storage':
+            content_url = file.storage.url(
+                file.name,
+                parameters={
+                    "ResponseContentType": full_headers.pop("Content-Type"),
+                    "ResponseContentDisposition": full_headers.pop("Content-Disposition"),
+                },
+            )
+            raise HTTPFound(content_url, headers=full_headers)
```

Comment:
I think 302 should not contain headers from full_headers at all.

Reply:
I agree, and i hope the tests do, too.